heartbeat/docs/autodiscover-kubernetes-config.asciidoc

```diff
@@ -0,0 +1,23 @@
+{beatname_uc} supports templates for modules:
+
+["source","yaml",subs="attributes"]
+-------------------------------------------------------------------------------------
+heartbeat.autodiscover:
+  providers:
+    - type: kubernetes
+      include_annotations: ["prometheus.io.scrape"]
+      templates:
+        - condition:
+            contains:
+              kubernetes.annotations.prometheus.io.scrape: "true"
+          config:
+            - type: http
+              hosts: ["${data.host}:${data.port}"]
+              schedule: "@every 1s"
+              timeout: 1s
+-------------------------------------------------------------------------------------
+
+This configuration launches a `http` module for all containers of pods annotated `prometheus.io.scrape=true`.
+There are cases where the PodSpec does not expose a port. In such cases the host can be provided as `${data.host}:9090`
```

Comment:
I agree that this is confusing. This language was borrowed from https://github.com/elastic/beats/blob/master/metricbeat/docs/autodiscover-kubernetes-config.asciidoc . Perhaps @exekias and @ruflin would either of you be able to help improve this text.

I'm not a kubernetes expert, so is the issue that by hard coding the port you could accidentally connect to the wrong container on the right host and get stats for that by mistake?

Reply:
@exekias I'm still confused by `metrics will not be enriched with container information because
the discovery mechanism won't be able to determine which container maps to the port.`. Perhaps this is not even necessary for heartbeat since we don't record metrics?

Reply:
you can switch `metrics` to `events`. This text is trying to explain that if the port is not exposed by the Pod, autodiscover won't know what container in the Pod it's talking to, so it will enrich only with Pod information, but not the container name.

Reply:
How's this copy sound then?

> In some cases container information may not be available, only pod information. If the PodSpec does not expose a port, you can specify the port directly in the `hosts` setting. For example: `${data.host}:9090`. However, events will not be enriched with container information.
>

Reply:
I'm not sure if people will realize that first statement (`In some cases container information may not be available, only pod information`) is caused by the next sentence, probably @dedemorton can give better input here than me

Reply:
I think that I am confused here as well then. @exekias I'm not a k8s expert. Is the port used to obtain container info from the pod? I thought the port was for the service running on the container. Is there some metadata service as well here? If that's the case we should say:

> If a port is not available it will not be possible to pull container info from the pod or to use the port for Heartbeat monitors. In this case you'll need to specify the port manually. For example: ${data.host}:9090. However, events will not be enriched with container information.

Reply:
So let me try to explain:

Pods are created as a group of containers. For each container, you can expose ports. When we go through the discovery process we spawn:

- An event for the entire pod, with the host=ip address of the pod and no port, no container name
- An event for every exposed port, including the container name

If you don't use the `${port}` variable you are using the first event, which doesn't contain info about the container exposing that port. In some cases this port is not defined in the kubernetes side, so we won't know about it.